### ISSUE TITLE
Support for bare RTP with H264

### ISSUE THREAD

**Issue by snachmsm** [google/ExoPlayer#10107]
I'm a bit lost with ExoPlayer support for bare RTP with H264. According to [DOC](https://exoplayer.dev/rtsp.html) there is a support for RTP/UDP with SDP, according to [THIS](https://github.com/google/ExoPlayer/issues/9104) issue - it's not... I've decided to give this player a try

So my SDP looks like below

```
v=0
o=- 0 0 IN IP4 null
s=Unnamed
i=N/A
c=IN IP4 192.168.137.1
t=0 0
a=recvonly
m=video 5006 RTP/AVP 96
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1;profile-level-id=42c029;sprop-parameter-sets=Z0LAKY1oB4AiflgHhEI1,aM4BqDXI;
a=control:trackID=1
```

There is a required `fmtp` field, profile/level/sps/pps are real ones, not hardcodes, produced by built-in Android `MediaCodec`. I can open this file in desktop VLC (Win/Lin) and everything works [almost](https://stackoverflow.com/questions/71527681/rtp-with-h264-glitched-depend-on-network-connection) fine. "Almost" so I've decided to try another player (QuickTime fails on desktop, greenscreen), preferably on Android.

I'm using current `2.17.1` version of core/ui/rtsp, loading SDP file as a... file

```
Uri uri = Uri.fromFile(file);
exoPlayer.setMediaItem(MediaItem.fromUri(uri));
exoPlayer.prepare();
exoPlayer.setPlayWhenReady(true);
```

and this gives me

```
ExoPlayerImplInternal: Playback error
  com.google.android.exoplayer2.ExoPlaybackException: Source error
    at com.google.android.exoplayer2.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:641)
    at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:611)
    at android.os.Handler.dispatchMessage(Handler.java:103)
    at android.os.Looper.loop(Looper.java:214)
    at android.os.HandlerThread.run(HandlerThread.java:67)
 Caused by: com.google.android.exoplayer2.source.UnrecognizedInputFormatException: None of the available extractors (FlvExtractor, FlacExtractor, WavExtractor, FragmentedMp4Extractor, Mp4Extractor, AmrExtractor, PsExtractor, OggExtractor, TsExtractor, MatroskaExtractor, AdtsExtractor, Ac3Extractor, Ac4Extractor, Mp3Extractor, JpegExtractor) could read the stream.
    at com.google.android.exoplayer2.source.BundledExtractorsAdapter.init(BundledExtractorsAdapter.java:92)
    at com.google.android.exoplayer2.source.ProgressiveMediaPeriod$ExtractingLoadable.load(ProgressiveMediaPeriod.java:1020)
....
```

Am I missing something or this bare RTP stream isn't supported (yet)? If it is supported (as in doc) how this should be initialized?

PS. Note that is NOT a RTSP, there is no control sequences, just pure UDPs pushed on some ip:port
PS2. [Screenshot from Wireshark](https://imgur.com/byY2jGn). Using some Lua script I can extract frames from stream and play them as raw H264, in VLC obviusly

**Comment by claincly** [comment]
We don't support side loading SDP at the moment, so yes, opening a SDP file (either local or remote) is not supported.

Can you explain why you need to side load this SDP? If you already can produce an RTP stream, it's likely you can get RTSP support with it.

**Comment by snachmsm** [comment]
Well, looks like RTSP isn't an option in MCVideo ([3GPP TS 26.281](https://www.etsi.org/deliver/etsi_ts/126200_126299/126281/14.00.00_60/ts_126281v140000p.pdf)). I'm working on similar-based project (critical communication) and in future it will migrate to this standard, but for now I'm just using pure RTP with own custom control (instead of RTCP required by MCVideo). I just don't want to use protocols, which can't be used in future migration.

My code is "very inspired" by [libstreaming](https://github.com/fyhertz/libstreaming) (not my import, got it like it is) - I'm using code similar to [example](https://github.com/fyhertz/libstreaming#how-to-stream-h264-and-aac) and there is a `getSessionDescription()` method - this string is shared through custom control to my server-side (which also receives stream directly). When any other client (receiver) will ask server-side for stream it will obtain a bit modified SDP (ip:port updated) and server-side will start make a copy received packets from streamer to receiver(s).

BTW.  there is an option for RTSP Server in `Service` in this lib, untouched in my import - VLC plays without problems, ExoPlayer after few secs of freeze shows one frame and freezes again till stream ends. Just tried few times, without any further investigation. I'm focusing on problem with [glitches](https://stackoverflow.com/questions/71527681/rtp-with-h264-glitched-depend-on-network-connection) rendered in VLC, just wanted to try another player.

**Comment by claincly** [comment]
I don't think I fully understand. It seems you want to cross reference VLC behaviour with another player? In that case I wouldn't suggest spend too much time hacking ExoPlayer to support sideloading SDP. And if VLC plays with glitches, it's likely there's something wrong with the streaming server.

**Comment by snachmsm** [comment]
I want to stream to my public server side and during this stream some other client may request a copy of this data. So there might be multiple receivers in fact (separated ports used), but each user may be in one state only: Preview, Streamer or Receiver (so handles at most 1 stream at once). This works, but I'm occuring glitches like on screen.

I've noticed that every way of video quality verification involves VLC - Android Receiver is using LibVLC, there is also another software - Win client - and it also uses .NET LibVLC, raw H264 data dumped from Wireshark is further opened in VLC desktop (Win/Lin verified), redirecting stream through local network straight to laptop - VLC desktop. I've just started suspecting VLC... But thats one and only soft which can open SDP-only RTP afaik, I'm just looking for some other way (on any OS in fact, but Android preferably).

I would also blame server, but that won't explain why glitches are present in laptops hotspot 137.100 -> 137.1 stream scenario, nor why they aren't present on some local Wi-Fi networks 19.100 -> 19.123 (without server side in the middle). With Wiresharks lua script I can dump outgoing stream and glitches are already there! But this raw dump is also opened in VLC...

There is no single glitch in `MediaCodec` output, I can pack NAL units into MP4 container, which is playable and without any issues. So now I would blame [packetization process](https://github.com/fyhertz/libstreaming/blob/master/src/net/majorkernelpanic/streaming/rtp/H264Packetizer.java), but this doesn't answer why it works well on some networks, on other not... And brings back to VLC suspection, as I'm clueless...

But that long story isn't related to ExoPlayer itself if it can't open raw SDP file. Would be nice if it could handle this in future release, just as a "simplest RTP case" (and also RTCP for e.g. MCVideo), but for now this issue may be closed I think... (but if you have any insight feel free to share ;) )

Thanks for support and this lib, is awesome
